### ISSUE TITLE
Validate multiple json instances with same schema

### ISSUE THREAD

**Issue by willolisp** [luposlip/json-schema#2]
Hi there,
first of, I'd like to thank you for all the work invested in this lib. It helped me a great deal in a personal project I've been working on. Thank you.
I'd also like to state a potential improvement point I came across while using the lib: validating json instances against a common json schema.
In my project, I have to validate considerably large batches of json instances (about 100k instances each batch) against a single, complex json schema.
The lib gets the job done, of course, but, since it must `prepare-schema` for each json instance, it ends up wasting a lot of time preparing the same schema time and time again.
It'd be great if the lib provided a function for validating multiple instances against a single schema.
By playing around with the lib, I managed to reduce the validation time from about 1min to about 1s just by not repeating the schema preparation step.
If you feel this would be a nice improvement to have, I can try and give it a look and maybe submit a PR. I'm admittedly very new to Clojure, but am willing to help any way I can :)
Cheers!

**Comment by luposlip** [comment]
Thanks for your input, makes a lot of sense.
I'll look into it this weekend.

fre. 10. maj 2019 00.05 skrev willolisp <notifications@github.com>:

**Comment by luposlip** [comment]
Fixed with https://github.com/luposlip/json-schema/releases/tag/0.1.5.
Closing issue.

You're welcome to reopen this if you find any issues.

**Comment by willolisp** [comment]
Hmmmm I believe this hasn't fixed it for me yet.

Maybe my use-case is too much of a specific one, but I'll try and explain it nonetheless.
I notice we're now able to validate json **arrays** against a schema without having to repeat the prepare-schema step, which is great! However, I'd like to be able to validate multiple json **objects** against a particular schema... The difference being that the schema should define a structure of type object, and not of type array.

This use-case may not make a lot of sense from a json-schema validation point of view, so I'll try to explain a little bit of the business here. The schemas are actually built by some end-user that wants to define the structure of an HTML form. Since these forms are stored as json objects in our DB, it makes sense for their schema to define objects, and not arrays.

So, basically, I'd like to have a way of calling `(.validate ^Schema prepared-schema prepared-json)` without having to prepare the schema every time.

Do you think this use-case is common enough that the lib should contemplate it? Or is it too specific to my business?

Cheers.

**Comment by luposlip** [comment]
Hi again,

I'm pretty sure this is what the test `validate-json` shows, or?

**Comment by willolisp** [comment]
Hmmmmm I may be missing something here, but both valid scenarios in the `validate-json` test case take a single json object (namely,` "{\"id\": 1}"` and `{:id 1}`) to be validated against a schema, no?
I'd like to be able to do something like this:
`(json/validate-many {:$schema "http://json-schema.org/draft-07/schema#" :type "object" ...} [{:field value}, ...])`. Notice the schema has root type of `object` and an array of jsons is supplied as targets for validation, so the `json/validate-many` function should validate each of the json objects in the array against the schema provided instead of validating the full array against said schema.
Maybe I'm not being clear enough (as I'm not a native english speaker), so let me know if you need an example or a more detailed explanation.
Cheers!

**Comment by luposlip** [comment]
For that I'd definitely use a pipeline. Read all docs in lazily, as to get
a `lazy-seq` of json documents to validate. Then validate them all, one by
one, as in the pipeline example in the readme.

Isn't that workable in your situation?

man. 13. maj 2019 23.42 skrev willolisp <notifications@github.com>:

**Comment by willolisp** [comment]
I was under the impression that that would cause the schema to be prepared multiple times, but, as I said before, I'm quite new to Clojure, so I'm probably wrong.
Will definitely give it a look! Thanks!
